lib/opentelemetry/context.lua

```diff
@@ -8,31 +10,88 @@ local mt = {
     __index = _M
 }
 
-local key = "opentelemetry-context"
+local context_key = "__opentelemetry_context__"
 
-function _M.new(storage)
-    return setmetatable({storage = storage, sp = noop_span, is_attached = false}, mt)
+
+--------------------------------------------------------------------------------
+-- Create new context with set of entries
+--
+-- @return              context
+--------------------------------------------------------------------------------
+function _M.new(entries)
+    return setmetatable({ sp = noop_span, entries = entries or {} }, mt)
 end
 
+--------------------------------------------------------------------------------
+-- Set this context as current by pushing it on stack stored at context_key.
+--
+-- @return              token to be used for detaching
+--------------------------------------------------------------------------------
 function _M.attach(self)
-    self.prev = self:current()
-    self.storage:set(key, self)
-    self.is_attached = true
+    if otel_global.context_storage[context_key] then
+        table.insert(otel_global.context_storage[context_key], self)
+    else
+        otel_global.context_storage[context_key] = { self }
+    end
+
+    -- the length of the stack is token used to detach context
+    return #otel_global.context_storage[context_key]
 end
 
-function _M.detach(self)
-    if self.is_attached then
-        self.storage:set(key, self.prev)
-        self.is_attached = false
+--------------------------------------------------------------------------------
+-- Detach current context, setting current context to previous element in stack
+-- If token does not match length of elements in stack, returns false and error
+-- string.
+--
+-- @return            boolean, string
+--------------------------------------------------------------------------------
+function _M.detach(self, token)
+    if #otel_global.context_storage[context_key] == token then
+        table.remove(otel_global.context_storage[context_key])
+        return true, nil
+    else
+        local error_message = "Token does not match (" ..
+            #otel_global.context_storage[context_key] .. " context entries in stack, token provided was " .. token .. ")."
+        ngx.log(ngx.WARN, error_message)
```

Comment:
Would be nice to have a centralized logger in the global module, but not necessary for this PR.